packages/dds/tree/src/feature-libraries/schemaIndex.ts

```diff
@@ -15,13 +15,168 @@ import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+        treeSchema.push(encodeTree(name, schema));
+    }
+    for (const [name, schema] of repo.globalFieldSchema) {
+        globalFieldSchema.push(encodeNamedField(name, schema));
+    }
+    treeSchema.sort(compareNamed);
+    globalFieldSchema.sort(compareNamed);
+    return {
+        version,
+        treeSchema,
+        globalFieldSchema,
+    };
+}
+
+function compareNamed(a: Named<string>, b: Named<string>) {
+    if (a.name < b.name) {
+      return -1;
+    }
+    if (a.name > b.name) {
+      return 1;
+    }
+    return 0;
+}
+
+function encodeTree(name: TreeSchemaIdentifier, schema: TreeSchema): TreeSchemaFormat {
+    const out: TreeSchemaFormat = {
+        name,
+        extraGlobalFields: schema.extraGlobalFields,
+        extraLocalFields: encodeField(schema.extraLocalFields),
+        globalFields: [...schema.globalFields].sort(),
+        localFields: [...schema.localFields].map(([k, v]) => encodeNamedField(k, v)).sort(compareNamed),
+        value: schema.value,
+    };
+    return out;
+}
+
+function encodeField(schema: FieldSchema): FieldSchemaFormat {
+    const out: FieldSchemaFormat = {
+        kind: schema.kind,
+    };
+    if (schema.types !== undefined) {
+        out.types = [...schema.types];
+    }
+    return out;
+}
+
+function encodeNamedField(name: string, schema: FieldSchema): NamedFieldSchemaFormat {
+    return {
+        ...encodeField(schema),
+        name,
+    };
+}
+
+function decode(f: Format): SchemaData {
+    const globalFieldSchema: Map<GlobalFieldKey, FieldSchema> = new Map();
+    const treeSchema: Map<TreeSchemaIdentifier, TreeSchema> = new Map();
+    for (const field of f.globalFieldSchema) {
+        globalFieldSchema.set(brand(field.name), decodeField(field));
+    }
+    for (const tree of f.treeSchema) {
+        treeSchema.set(brand(tree.name), decodeTree(tree));
+    }
+    return {
+        globalFieldSchema,
+        treeSchema,
+    };
+}
+
+function decodeField(schema: FieldSchemaFormat): FieldSchema {
+    const out: FieldSchema = {
+        kind: schema.kind,
+        types: schema.types === undefined ? undefined : new Set(schema.types),
+    };
+    return out;
+}
+
+function decodeTree(schema: TreeSchemaFormat): TreeSchema {
+    const out: TreeSchema = {
+        extraGlobalFields: schema.extraGlobalFields,
+        extraLocalFields: decodeField(schema.extraLocalFields),
+        globalFields: new Set(schema.globalFields),
+        localFields: new Map(
+            schema.localFields.map((field): [LocalFieldKey, FieldSchema] => [brand(field.name), decodeField(field)]),
+        ),
+        value: schema.value,
+    };
+    return out;
+}
+
+/**
+ * Synchronous monolithic summarization of schema content.
+ *
+ * TODO: when perf matters, this should be replaced with a chunked async version using a binary format.
+ *
+ * @returns a snapshot of the schema as a string.
+ */
+export function getSchemaString(data: SchemaData): string {
```

Comment:
Do you want these to be exported?

We could also consider moving most of these types and helpers to a separate module which SchemaIndex imports.

Reply:
These are tested, so they have to be exported. They are not exported from the parent module or package.

As these define the persisted format of SchemaIndex, they really belong as part of it. We could split this file and make a folder for both files, but I don't think that offers much value other than making these test exports slightly smaller scoped.